Users can change ownership of their video [#510]

grizio opened the pull request:
As described in #510, users can change ownership of their video.

Please look at the list of commits that explain how it was implemented.

# QA

## Prerequisites

* You need two accounts
* You need at least one video in one account

## Give the ownership

* Go to « My videos »
* Click on button « Change ownership »
* A modal opens
* Write the target user name (autocompletion should work starting with one character)
* Validate
* The request is sent

## Accept the ownership

* Change account to use the one of the targeted user from first scenario
* Go to « My account »
* Go to « Ownership change »
* You see the request created in first scenario
* Click on « Accept »
* A modal opens
* Select the channel
* Validate
* If you go to the list of your videos, you will see the new video

## Refuse the ownership

* Change account to use the one of the targeted user from first scenario
* Go to « My account »
* Go to « Ownership change »
* You see the request created in first scenario
* Click on « Refuse »
* If you go to the list of your videos, you will not see the new video

## Other things to test

* You cannot request the change of an ownership for another user
* You cannot accept or refuse the ownership for another user

## Gif time

## Scenario 1

![accept](https://user-images.githubusercontent.com/11132978/44626707-7367d680-a921-11e8-92ba-0bd91ea72f63.gif)

* User 1 : ask the change of ownership to user 2
* User 2 : accept the change of ownership

## Scenario 2

![refuse](https://user-images.githubusercontent.com/11132978/44626711-7cf13e80-a921-11e8-9411-14df7471fc26.gif)

* User 1 : ask the change of ownership to user 2
* User 2 : refuse the change of ownership

# TODO List

- [x] Create the main feature end-to-end
- [x] Move the ownership change to my video page
- [x] Create icons for:
  - [x] Change ownership button (:warning: Currently report abuse button!)
  - [x] Accept / refuse the ownership (none set)
- [x] Add specific right (https://github.com/Chocobozzz/PeerTube/pull/888#discussion_r206417754)
- [x] Add automatic tests
- [x] Add controls to avoid sending twice the same request
- [ ] Correct the ActivityPub part (https://github.com/Chocobozzz/PeerTube/pull/888#discussion_r206417337)
- [x] Add controls checking user quotas when accepting a request
- [x] Check i18n
- [ ] Check the user experience of the whole process (should need a designer)
- ~Add notifications (See #493, could be reported?)~

rigelk commented:
Could you post user stories with pictures to allow first-hand review by designers, so that they don't have to run the PR code to do basic tests?

rigelk commented on client/src/app/videos/+video-watch/modal/video-change-ownership.component.ts:
```diff
@@ -0,0 +1,77 @@
+  }
+
+  search (event) {
+    const query = event.query;
```
unnecessary semicolon making the linter cry.

Chocobozzz commented:
Looks good!

> The request disappears from the list (maybe could need a confirmation?)

I think just marking this *Refused* would be better (history purpose etc) instead of deleting it

Chocobozzz commented on client/src/app/+my-account/my-account-ownership/my-account-ownership.component.ts:
```diff
@@ -0,0 +1,73 @@
+      )
+  }
+
+  createByString = (account: Account) => {
```
Juste create classic method:

```
createByString (account: Account) {

}
```

Chocobozzz commented on server/controllers/api/videos/ownership.ts:
```diff
@@ -0,0 +1,108 @@
+  asyncRetryTransactionMiddleware(listVideoOwnership)
+)
+
+ownershipVideoRouter.post('/ownership/:id/accept',
```
I prefer this `post` instead of a REST `put`, because it's an irreversible action so :+1:

Chocobozzz commented on server/controllers/api/videos/ownership.ts:
```diff
@@ -0,0 +1,108 @@
+)
+
+ownershipVideoRouter.get('/ownership',
+  authenticate,
```
Use 

```
  paginationValidator,
  setDefaultPagination,
```

middlewares for pagination

grizio replied:
Done (not marked as outdated)

Chocobozzz commented on server/controllers/api/videos/ownership.ts:
```diff
@@ -0,0 +1,108 @@
+  const body: VideoChangeOwnershipCreate = req.body
+
+  await sequelizeTypescript.transaction(async t => {
+    const nextOwner = await AccountModel.findOne({ where: { name: body.username } })
```
Use `loadLocalByName` from `AccountModel`

Chocobozzz commented on server/controllers/api/videos/ownership.ts:
```diff
@@ -0,0 +1,108 @@
+    const targetVideo = videoChangeOwnership.Video
+    const channel = res.locals.videoChannel as VideoChannelModel
+
+    targetVideo.set('channelId', channel.id)
```
Now comes the ActivityPub parts. 
 * Send video update. This might not work for now because in the `processUpdate` function we check the actor owns the video channel. I think we should send the activity from the old actor (so the activity is signed with the old owner) and in the `processUpdate` function, check the `attributedTo` `Person` owns the video channel (instead of the `videoChannel.Account.Actor.id !== actor.id` condition). Send me en email if you need more help
 * Use `changeVideoChannelShare` method

Chocobozzz commented on server/middlewares/validators/videos.ts:
```diff
@@ -221,6 +223,47 @@ const videosShareValidator = [
 ]
 
+const videosChangeOwnershipValidator = [
+  param('id').custom(isIdOrUUIDValid).not().isEmpty().withMessage('Should have a valid id'),
```
I don't think this id could be an UUID?

grizio replied:
It is the video id. I used the same validations than other functions (like `videosUpdateValidator` in same file).

Chocobozzz replied:
Oh right, maybe `videoId` would be a better parameter name

Chocobozzz commented on server/middlewares/validators/videos.ts:
```diff
@@ -221,6 +223,47 @@ const videosShareValidator = [
+    if (!await isVideoExist(req.params.id, res)) return
+
+    // Check if the user who did the request is able to change the ownership of the video
+    if (!checkUserCanManageVideo(res.locals.oauth.token.User, res.locals.video, UserRight.UPDATE_ANY_VIDEO, res)) return
```
Maybe create a specific right would be better? ping @rigelk

rigelk replied:
Yes, I think we should create another right.

Chocobozzz commented on server/models/account/user.ts:
```diff
@@ -303,4 +303,15 @@ export class UserModel extends Model<UserModel> {
+    return UserModel.findAll({
+      where: {
+        username: {
+          [Sequelize.Op.like]: `%${search || ''}%`
```
`search` should be already checked so we should not have to do the `||`

Chocobozzz commented on server/models/video/video-change-ownership.ts:
```diff
@@ -0,0 +1,132 @@
+        nextOwnerAccountId: nextOwnerId
+      },
+      include: [
+        {
```
You could create a scope for these includes

Chocobozzz commented on server/helpers/custom-validators/video-ownership.ts:
```diff
@@ -0,0 +1,42 @@
+): boolean {
+  if (videoChangeOwnership.NextOwner.userId === user.Account.userId) {
+    return true
+  } else {
```
Don't need this `else` because you `return true`

grizio replied:
Hmm… to better understand the algorithm, we should keep the else part. It gives a quick indication of the function structure (code pattern).

Same for other comments below.

Chocobozzz replied:
The else is unneeded, it's implicit because the `return` is just above. It adds 2 more lines and makes it more difficult to see the important information: the `return`.

They are two different visions but to keep consistency with the rest of the project, please remove it

Chocobozzz commented on server/helpers/custom-validators/video-ownership.ts:
```diff
@@ -0,0 +1,42 @@
+async function loadVideoChangeOwnership (id: string): Promise<VideoChangeOwnershipModel | undefined> {
+  if (validator.isInt(id)) {
+    return VideoChangeOwnershipModel.load(parseInt(id, 10))
+  } else {
```
Don't need the `else`

Chocobozzz commented on server/controllers/api/videos/ownership.ts:
```diff
@@ -0,0 +1,108 @@
+
+      logger.info('Ownership change for video %s created.', videoInstance.name)
+      return res.type('json').status(204).end()
+    } else {
```
Don't need the `else`

Chocobozzz commented on server/controllers/api/users.ts:
```diff
@@ -74,6 +74,10 @@ usersRouter.get('/me/videos/:videoId/rating',
 )
 
+usersRouter.get('/autocomplete',
+  asyncMiddleware(autocompleteUsers)
```
Add a middleware to check the params of this route (so that search exists)

Chocobozzz commented:
It's an awesome PR @grizio, congrats! 

I'm just wondering if you should not put the give ownership modal/button in the **my account/videos** tab instead of the watch page? (With maybe a shortcut in the watch page).

grizio commented:
Updated the PR with your comments, I split commits to be able to squash afterward.

Also updated TODO list for remaining comments.

> I'm just wondering if you should not put the give ownership modal/button in the my account/videos tab instead of the watch page? (With maybe a shortcut in the watch page).

I am waiting feedback from designers before doing this update to avoid changing it too much.

rigelk commented:
I would also agree with @Chocobozzz on moving the action and modal on the *my account/videos* tab. It could come useful once we want to support batch actions.

rigelk commented:
@grizio concerning the icon to use, I think this one should be fine: 

![to_other](https://user-images.githubusercontent.com/6329880/44402169-37afc400-a551-11e8-8110-c0334ef94903.png)

It's the "I'm with her" icon from https://robbiepearce.com/softies/ where we already take our icons.

Otherwise there is the 'users' icon, which is maybe less abstract.

Chocobozzz commented on client/src/app/+my-account/my-account-ownership/my-account-ownership.component.html:
```diff
@@ -0,0 +1,54 @@
+      <td>{{ videoChangeOwnership.createdAt }}</td>
+      <td i18n>{{ videoChangeOwnership.status }}</td>
+      <td class="action-cell">
+        <ng-template [ngIf]="videoChangeOwnership.status === 'WAITING'">
```
Prefer using `<ng-container *ngIf=...`

Chocobozzz commented on client/src/app/+my-account/my-account-ownership/my-account-accept-ownership/my-account-accept-ownership.component.html:
```diff
@@ -0,0 +1,34 @@
+  <div class="modal-body" [formGroup]="form">
+    <div class="form-group">
+      <label i18n for="channel">Select the target channel</label>
+      <select formControlName="channel" id="channel">
```
Use `.peertube-select-container`

Chocobozzz commented on client/src/app/+my-account/my-account-videos/video-change-ownership/video-change-ownership.component.ts:
```diff
@@ -0,0 +1,75 @@
+      .open(this.modal)
+      .result
+      .then(() => this.changeOwnership())
+      .catch((_) => _)
```
You could log the error in `console.error` at least

Chocobozzz replied:
Oki! Could you add a small comment about this in the code please?

Chocobozzz commented on client/src/app/shared/buttons/button.component.html:
```diff
@@ -0,0 +1,4 @@
+<span [class]="getClassName()" [title]="label">
```
```<span class="icon" [ngClass]="className" [title]="label">```

Chocobozzz commented on client/src/app/shared/buttons/button.component.html:
```diff
@@ -0,0 +1,4 @@
+<span [class]="getClassName()" [title]="label">
+  <span [class]="getIconName()"></span>
```
And same here

Chocobozzz commented on server/controllers/api/videos/ownership.ts:
```diff
@@ -0,0 +1,119 @@
+  const body: VideoChangeOwnershipCreate = req.body
+
+  await sequelizeTypescript.transaction(async t => {
+    const nextOwner = await AccountModel.loadLocalByName(body.username)
```
`nextOwner` should be fetched in the validator middleware

Chocobozzz commented on server/controllers/api/videos/ownership.ts:
```diff
@@ -0,0 +1,119 @@
+    const nextOwner = await AccountModel.loadLocalByName(body.username)
+    if (!nextOwner) return res.type('json').status(400).end()
+
+    const existingRequest = await VideoChangeOwnershipModel.findExisting(initiatorAccount.id, nextOwner.id, videoInstance.id)
```
You could use `findOrCreate` sequelize method

Chocobozzz commented on server/models/video/video-change-ownership.ts:
```diff
@@ -0,0 +1,140 @@
+  updatedAt: Date
+
+  @Column
+  status: VideoChangeOwnershipStatus
```
Add `@AllowNull(false)`

Chocobozzz commented on server/models/video/video-change-ownership.ts:
```diff
@@ -0,0 +1,140 @@
+  }
+
+  static load (id: number) {
+    return VideoChangeOwnershipModel.scope(ScopeNames.FULL).find({
```
`findById`

Chocobozzz commented on server/models/video/video-change-ownership.ts:
```diff
@@ -0,0 +1,140 @@
+  }
+
+  static findExisting(initiatorAccountId: number, nextOwnerAccountId: number, videoId: number) {
+    return VideoChangeOwnershipModel.find({
```
`findOne` (consistency + more explicit)

Chocobozzz commented:
@grizio Thanks! I just have a few remarks, but LGTM. For the federation I can do it after the merge. Tell me when it is okay for you :)

rigelk commented on server/models/video/video-change-ownership.ts:
```diff
@@ -0,0 +1,140 @@
+    })
+  }
+
+  static findExisting(initiatorAccountId: number, nextOwnerAccountId: number, videoId: number) {
```
there is a missing whitespace after the function name

fflorent commented:
Just a comment to give some feedback: This feature would just be awesome! Thanks for working on it! <3

Florent

grizio commented:
@Chocobozzz I made these last changes and merge with develop.

I tried to clean my history but because of merges, I cannot do it without triggering the merge another time and risking to add bugs.

Chocobozzz commented:
No problem, I squashed your commits. Thanks for your great work @grizio!

fflorent commented:
I look forward to trying it. Thanks a lot @grizio! 

Florent

plslvval commented:
Is it still working in 1.30/1.3.1 ?
I don't find  button « Change ownership » in « My videos » ( remark my environment is french so wording is different ).

Chocobozzz commented:
@plslvval Yes it should work in 1.3.1. It's *Changer de propriétaire* in french. If you don't see it, try with another web browser or disable your web extensions.

plslvval commented:
@Chocobozzz 
thanks, i tested from home and it does work, so this is something related to the system i did first test with.
